File: ChartTest.php
<!DOCTYPE html>
<html>
<head>
	<!-- 
		Author: John Bartel
		Last Mod: Feb-14 2019
	 -->
	<meta charset="utf-8">
	<style type="text/css">
		body {background-color: lightgrey; margin: 0 2%;}
		#fraym {width: 100%;}
	</style>
	<?php
		//Connection settings
		$DBHost = 'localhost';
		$DBUser = 'root';
		$DBPass = '';
		$DBName = 'gps';
		$DBTable = 'test';

		$readings = array();											//Array to hold points for graph.
		if($conn = mysqli_connect($DBHost, $DBUser, $DBPass, $DBName)){	//Set connection, proceed if successful.
			$qry = "SELECT `trace_data` as `trace_data`, `trace_time` FROM $DBTable ORDER BY `trace_id`";
			if($result = mysqli_query($conn, $qry)){
				while($data=mysqli_fetch_assoc($result)){
					$temp = bin2hex(iconv('UTF-8', 'UCS-2BE', $data['trace_data'])); 	//Converts result to 16-bit hex.
					$start = $end = 0;												//Position vars.
					$points = array();
					for($i = 0; $i < strlen($temp)/16; $i++){
						$tempStr1 = substr($temp, $start, 16);						//Takes 16 hex characters for number.
						$start+=16;
						$tempStr2 = '';
						$slice = 0;
						for($j = 0; $j < strlen($tempStr1); $j++){
							$tempStr2 .= substr(substr($tempStr1, $slice, 4), 2);	//For each group of 4 take last 2 since we need only last 8 bits since first 8 are 00.
							$slice += 4;
						}
						$points[] = unpack("l", pack("l", hexdec("0x".$tempStr2)))[1]/1000;	//Converts to signed int required for graphing.
						//print_r($points);
					}
					$readings[]=array($points, $data['trace_time']);
				}
			}
			mysqli_close($conn);
		}else{ 							//Else kill script.
		  die("Connection error: Please Contact an Administrator");
		}
	?>
	<script type="text/javascript">
		function createGraph(){
			var readings = <?php echo json_encode($readings); ?>;	//encoded to json and set as javascript var.
			var svgns = "http://www.w3.org/2000/svg";				//Namespace for svg.
			var chart = document.createElementNS(svgns, "svg:svg"); //Create svg in namespace.
			var index = 0;				//Index of readings array for values.  Modulus by array length so it repeats on completion.
			var xNext = xLine = 0; 			//starting points
			var vOffset = 50;								//Vertical offset for plotting points.	
			var coorStr = '';										//Coordinate string for graph points.
			const hAxisL = 850;										//Horizontal axis ticks lower value.
			const hAxisU = 1150;									//Horizontal axis ticks upper value.
			const hAxisDiv = 2;										//Horizontal axis division count.
			const hAxisInt = (hAxisU-hAxisL)/hAxisDiv;				//Horizontal axis tick interval amount.
			var hAxisSep = 0;										//Tracks horizontal axis tick count.

			for(var i = 0; i < readings[index][0].length; i++){
				var point = readings[index][0][i];					//Sets y axiz point from array.
		    	coorStr += xNext+","+(0-point)+" ";			//Saves to coordinate string to be used to draw line.
				xNext = xLine++;									//Increments xaxis position points.
			}
			chart.appendChild(createPolyline(coorStr, 0.3, "yellow"));	//Add line to chart.
			xLine-=2;												//Have to subtract 2 since it ends 2 points farther then last plotted line.

			chart.setAttribute("viewBox", "0 0 " + xLine + " " + vOffset);											//Sets view box size
			chart.setAttribute("style", "background-color: black;");
			document.getElementById("fraym").src ="data:image/svg+xml;charset=utf-8,"+(new XMLSerializer).serializeToString(chart);  		//Serializes svg so it can be set as image src.
		}

		function createPolyline(coord,s,c){	//Returns lines svg element. Takes line coordinates, stroke, and line color arguments relatively.
			var svgns = "http://www.w3.org/2000/svg";
			var line = document.createElementNS(svgns, "polyline");
			line.setAttribute("points",coord);
			line.setAttribute("stroke", c);
			line.setAttribute("stroke-width", s);
			line.setAttribute("fill", "none");
			return line;
		}

		//Sets interval for chart to call once a second.
		if (window.addEventListener) {
		   window.addEventListener("load", createGraph, false);
		} else if (window.attachEvent) {
		   window.attachEvent("onload", createGraph);
		}
	</script>
	<title>Chart Test</title>
</head>
<body>
	<h1>Chart:</h1>
	<img id='fraym' />
</body>
</html>

File: ChartTest.py
##  Author: John Bartel
##  Last Mod: Feb-14 2019

import binascii #For decoding the blob.
import struct #For converting hex string to signed int.
import time   #For setting interval of chart calls.

##External dependencies
import mysql.connector  #For Connecting to DB.
import matplotlib.pyplot as plt   #For plotting the points.
import matplotlib.animation as animation  #Setting a interval to change plot.

#Database connections settings
mydb = mysql.connector.connect(
  host="localhost",
  user="root",
  passwd="",
  database="gps"
)
sql = "SELECT trace_id, trace_data, trace_time FROM test ORDER BY trace_id"
tableCall = mydb.cursor() #Connection var.
tableCall.execute(sql)
result = tableCall.fetchall() #Var to hold database query result.

figure = plt.figure() #Create a figure for plots.
chart = figure.add_subplot(1,1,1) #Adds generic subplot that can be cleared and changed.
counter = 0 

def createGraph(i):
    global counter
    index = counter%len(result)
    counter += 1  #counter for getting values from query. Mod by length
    temp = binascii.hexlify(bytes(result[index][1].decode('utf-8'), 'utf-16-BE')).decode('utf-8') #Converts the block to a hex string. 
    point = [[],[]]   #coordinate array.
    xPos = 850    #Starting x pox so it can start at 850 MHZ.

    #Slices string into 16 piece strings because during conversion each bin char was converted to 8-bit hex.
    for i in range(0, len(temp)+1, 16):
        tempHex = ""
        tempSub = temp[(i-16):i]

        #Takes last 2 chars of string slices and combines.
        for j in range(0, len(tempSub)+1, 4):
            tempHex += tempSub[(j-4):j][2:4]
        #Have to check for null chars.
        if tempHex == '':
            continue
        else: #Converts hex string to signed int and divide by 1k.
            yPos = struct.unpack('>i', bytes.fromhex(tempHex))[0]/1000
            #Appends coordinates to array.
            point[0].append(xPos)
            point[1].append(yPos)
            xPos+=0.5

    chart.clear()   #Clear plot between iterations.
    chart.plot(point[0], point[1], color = "yellow", linewidth = 0.7) #plots the graph.
    chart.axis([850, 1150, -60, -30])
    chart.set_xlabel(result[index][2])    #Date stamp along xaxis.
    chart.set_yticks([-60, -50, -40, -30])
    chart.set_yticklabels(['-60 dBm', '-50 dBm','-40 dBm','-30 dBm'])
    chart.set_xticks([850,1000,1150])
    chart.set_xticklabels(['850 MHZ','1000 MHZ','1150 MHZ'])
    chart.set_title("Trace ID: "+str(result[index][0]))   #trace_id along y.
    chart.grid(linestyle='--', linewidth='0.4', color='grey') #inner grid marks.
    chart.set_facecolor('black')
#Sets the animation interval for 1 second and calls createGraph.
graphStep = animation.FuncAnimation(figure, createGraph, interval=1000)
plt.show()
